accepted/future-releases/0546-patterns/exhaustiveness.md

```diff
@@ -246,772 +251,854 @@ also be an instance of one of the known set of subtypes. That gives us the
-    If `changed` is empty, then no fields come into play and the result is
-    simply `L`.
+    1.  Let `n` be `headMax + tailMax`. *The challenge with list patterns is
+        that we could expand them to an unbounded number of spaces: empty list,
+        list with one element, list with two elements, etc. Here, `n` represents
+        the highest individual arity we need to expand to because there is no
+        pattern that will match a list with a specific number of elements larger
+        than that.*
 
-3.  Otherwise, the result is a union of objects where each object includes one
-    of the changed fields and leaves the others alone. This is how we tile over
-    the holes created by subtracting `R`.
+    2.  For `i` from `0` to `n`, half-inclusive:
 
-    1.  For each field in `changed`, create an object with the same type as `L`
-        with all fields from `fixed`, this field from `changed` and all other
-        field names in `changed` set to their original space in `L`.
+        1.  A space with type `T` and an arity restriction with length `i` and
+            `hasRest` `false`.
 
-    2.  The result is the union of those spaces.
+    3.  And a space with type `T` and an arity restriction with length `n` and
+        `hasRest` `true`.
 
-    For example:
+    *For example, given:*
 
-    ```
-    (w: bool, x: bool, y: bool, z: bool) - (x: true, y: false, z: true)`
+    ```dart
+    switch (list) {
+      case [_, ..., _, _]:
+      case [_, _, _, ..., _]:
+    }
     ```
 
-    `fixed` is `(w: bool)`. `changed` is `(x: false, y: true, z: false)`. The
-    result is a union of:
+    *The expansion of the value list is:*
 
     ```
-    (x: false, y: bool, z: bool) // Set x to D.
-    (x: bool, y: true, z: bool) // Set y to D.
-    (x: bool, y: bool, z: false) // Set z to D.
+    List(length: 0, hasRest: false)
+    List(length: 1, hasRest: false)
+    List(length: 2, hasRest: false)
+    List(length: 3, hasRest: false)
+    List(length: 4, hasRest: false)
+    List(length: 4, hasRest: true)
     ```
 
-## Space intersection
+*   Otherwise, `s` does not expand and the result is just `s`.
 
-Intersection is (mercifully) simpler than subtraction. In fact, we don't even
-need to calculate the actual intersection. We just need to know if it's empty
-or not.
+The expansion procedure is applied recursively to each resulting space described
+here until no more expansions take place.
 
-To calculate if the intersection `I` of spaces `L` and `R` is empty:
+*For example, the expansion of `FutureOr<bool?>` is:*
 
-1.  If `L` or `R` is empty, then `I` is empty.
+```
+Future<bool?>
+true
+false
+null
+```
 
-2.  If `L` or `R` is a union, then `I` is empty if the intersection of every
-    arm of the union with the other operand is empty.
+### Overapproximation
 
-3.  Otherwise, we're intersecting two object spaces.
+The *overapproximation* of a type `T` is `T` with all type variables replaced
+with their *defaults*:
```

Comment:
I think this is slightly underdefined.

I think we only apply "overapproximation" to `class`, `mixin` and `enum` declarations.  So say that.
(It gets weird if we try to replace type variables of a generic function type.)
We can always extend it to apply to the type of such a declaration by going from the type to the declaration.

Also, we should probably say that we apply it to the *declaration*, not the type.
(Types do not have type variables or bounds, their declarations - and, depending on mood, their generic classes and interfaces, which are still different from the *types* - do. The types have type *arguments*, but the variables are long gone.)

So, given a generic enum, mixin or class declaration, instantiate its type with the default types for its declaration's type parameters.
But that's basically what "instantiate to bounds" (I2B) does, giving you a valid super-bounded supertype for all valid instantiations, so maybe just use that?
(I2B also accounts for bound expressions that aren't straight-up types, up-to and including F-bounded type parameters. Can't just use a bound as a type directly.)

Reply:
I tend to think of this as "recursively instantiate to bounds" and it is slightly different that I2B. I2B works on raw types whereas the overapproximation works on instantiated types.

For instance the overapproximation of `Map<K, Map<V, K Function(K)>>>` in:
```
method<K extends num, V extends Foo<V>>(Map<K, Map<V, K Function(K)>>> map) => switch (map) { ... };
```
is `Map<num, Map<dynamic, num Function(Never)>>>`, so that each type variable occurring in the type has been replaced with the I2B default type modulo variance.

Reply:
Reworded to make it clearer that it applies to declarations.

Reply:
I'm not sure I follow it either, but that could be because I don't actually *understand* I2B.
Or what is being attempted here.

So let me try again:

Given a *static type* (which may contain free type variables referencing surrounding declarations),
take each such occurrence of a free type variable in the type, and replace it with its default type which is:
* If the referenced type variable has a bound, the default is the overapproximation of that bound.
* Otherwise if the free type variable occurs covariantly, the default is `Object?`.
* Otherwise the default is `Never`

The resulting type has no free type variables and will be a valid super-bounded type.

... But it's not valid super-bounded. If the variable has a bound with other free type variables, the first variable occurs contravariantly, we need an *underapproximation* to use instead.

And what if it occurs invariantly, like `List<void Function<X extends T>(X)>`?


I think it can make sense to talk about type variables in static types (I'm just not used to doing so), but I think here are more details to handle, if the goal is to get a valid super-bounded type as result.